lib_eio_linux/getaddrinfo_stubs.c

```diff
@@ -0,0 +1,187 @@
+
+#ifndef EAI_ADDRFAMILY
+#define EAI_ADDRFAMILY (-3000)
+#endif /* EAI_ADDRFAMILY */
+
+#ifndef EAI_BADHINTS
+#define EAI_BADHINTS (-3013)
+#endif /* EAI_BADHINTS */
+
+#ifndef EAI_NODATA
+#define EAI_NODATA (-3007)
+#endif /* EAI_NODATA */
+
+#ifndef EAI_OVERFLOW
+#define EAI_OVERFLOW (-3009)
+#endif /* EAI_OVERFLOW */
+
+#ifndef EAI_PROTOCOL
+#define EAI_PROTOCOL (-3014)
+#endif /* EAI_PROTOCOL */
+
+static int gai_errors[] = {
+  EAI_ADDRFAMILY,
+  EAI_AGAIN,
+  EAI_BADFLAGS,
+  EAI_BADHINTS,
+  EAI_FAIL,
+  EAI_FAMILY,
+  EAI_MEMORY,
+  EAI_NODATA,
+  EAI_NONAME,
+  EAI_OVERFLOW,
+  EAI_PROTOCOL,
+  EAI_SERVICE,
+  EAI_SOCKTYPE,
+  EAI_SYSTEM /* NOTE: must be last */
+};
+
+#define nmemb_gai_errors (sizeof(gai_errors) / sizeof(int))
+
+CAMLprim value caml_eio_getaddrinfo(value vnode, value vserv, value vopts)
+{
+  CAMLparam3(vnode, vserv, vopts);
+  CAMLlocal3(vres, v, vret);
+  char * node, * serv;
+  struct addrinfo hints;
+  struct addrinfo * res, * r;
+  int retcode, i;
+
+  if (! (caml_string_is_c_safe(vnode) && caml_string_is_c_safe(vserv)))
+    CAMLreturn (Val_emptylist);
+
+  /* Extract "node" parameter */
+  if (caml_string_length(vnode) == 0) {
+    node = NULL;
+  } else {
+    node = caml_stat_strdup(String_val(vnode));
+  }
+  /* Extract "service" parameter */
+  if (caml_string_length(vserv) == 0) {
+    serv = NULL;
+  } else {
+    serv = caml_stat_strdup(String_val(vserv));
+  }
+  /* Parse options, set hints */
+  memset(&hints, 0, sizeof(hints));
+  hints.ai_family = PF_UNSPEC;
+  for (/*nothing*/; vopts != Val_emptylist; vopts = Field(vopts, 1)) {
+    v = Field(vopts, 0);
+    if (Is_block(v))
+      switch (Tag_val(v)) {
+      case 0:                   /* AI_FAMILY of socket_domain */
+        hints.ai_family = caml_unix_socket_domain_table[Int_val(Field(v, 0))];
+        break;
+      case 1:                   /* AI_SOCKTYPE of socket_type */
+        hints.ai_socktype = caml_unix_socket_type_table[Int_val(Field(v, 0))];
+        break;
+      case 2:                   /* AI_PROTOCOL of int */
+        hints.ai_protocol = Int_val(Field(v, 0));
+        break;
+      }
+    else
+      switch (Int_val(v)) {
+      case 0:                   /* AI_NUMERICHOST */
+        hints.ai_flags |= AI_NUMERICHOST; break;
+      case 1:                   /* AI_CANONNAME */
+        hints.ai_flags |= AI_CANONNAME; break;
+      case 2:                   /* AI_PASSIVE */
+        hints.ai_flags |= AI_PASSIVE; break;
+      }
+  }
+  /* Do the call */
+  caml_enter_blocking_section();
+  retcode = getaddrinfo(node, serv, &hints, &res);
+  caml_leave_blocking_section();
+  if (node != NULL) caml_stat_free(node);
+  if (serv != NULL) caml_stat_free(serv);
+  /* Convert result */
+  vres = Val_emptylist;
+  if (retcode == 0) {
+    for (r = res; r != NULL; r = r->ai_next) {
+      v = caml_alloc_small(2, Tag_cons);
+      Field(v, 0) = convert_addrinfo(r);
```

Comment:
This violates GC rule 5 - the `convert_addrinfo` call needs to be done before the `caml_alloc_small` and stored in a registered local (`convert_addrinfo` allocates).

Reply:
thanks for spotting this, fixed it now